src/System.Net.Security/tests/FunctionalTests/SslStreamSniTest.cs

```diff
@@ -51,6 +51,132 @@ public void SslStream_ClientSendsSNIServerReceives_Ok(string hostName)
                 });
         }
 
+        [Theory]
+        [MemberData(nameof(HostNameData))]
+        public void SslStream_ServerCallbackAndLocalCertificateSelectionSet_Throws(string hostName)
+        {
+            X509Certificate serverCert = Configuration.Certificates.GetSelfSignedServerCertificate();
+
+            int timesCallbackCalled = 0;
+
+            var selectionCallback = new LocalCertificateSelectionCallback((object sender, string targetHost, X509CertificateCollection localCertificates, X509Certificate remoteCertificate, string[] issuers) =>
+            {
+                Assert.True(false, "LocalCertificateSelectionCallback called when AuthenticateAsServerAsync was expected to fail.");
+                return null;
+            });
+
+            var validationCallback = new RemoteCertificateValidationCallback((object sender, X509Certificate certificate, X509Chain chain, SslPolicyErrors sslPolicyErrors) =>
+            {
+                Assert.Equal(serverCert, certificate);
+                return true; 
+            });
+
+            VirtualNetwork vn = new VirtualNetwork();
+            using (VirtualNetworkStream serverStream = new VirtualNetworkStream(vn, isServer: true),
+                                        clientStream = new VirtualNetworkStream(vn, isServer: false))
+            using (SslStream server = new SslStream(serverStream, false, null, selectionCallback),
+                             client = new SslStream(clientStream, leaveInnerStreamOpen: false, validationCallback))
+            {
+                Task clientJob = Task.Run(() => {
+                    client.AuthenticateAsClient(hostName);
+                    Assert.True(false, "RemoteCertificateValidationCallback called when AuthenticateAsServerAsync was expected to fail.");
+                });
```

Comment:
I'm a little worried about this test randomly failing due to the inherent race condition you are setting up here by using Task.Run() to start the client-side of the test.  You're assuming that this will run immediately/concurrently when the latter AuthenticateAsServerAsync code runs.  Do you think this will be a problem?

Reply:
It's worth validating. I think the virtual connection and also regular sockets will wait for data and this will always work correctly but it makes sense to verify if that's the case (i.e. run 1000 threads doing that in a loop for 30s) - if it fails we might want to check my tests as well :)